### ISSUE TITLE
Calling a super constructor with an @required parameter that is not defaulted

### ISSUE THREAD

**Issue by dnfield** [dart-lang/linter#2008]
BAD:

```dart
class Foo {
  Foo(@required this.foo): assert(foo != null);

  int foo;
}

class Bar extends Foo {
  Bar({int foo}) : super(foo);
}
```

GOOD:

```dart
class Foo {
  Foo(@required this.foo): assert(foo != null);

  int foo;
}

class Bar extends Foo {
  Bar({@required int foo}) : super(foo);
}
```

I suspect with NNBD such a lint will be much less important (I'm not even sure what `@required` really means there anymore to be honest).  See https://github.com/flutter/flutter/issues/51363 or more context.

**Comment by bwilkerson** [comment]
Under Null Safety (formerly known as NNBD), the `@required` annotation has been formalized as a `required` keyword. The keyword has the same semantics except that it is possible to produce an error in cases (such as the one in the constructor for `Bar` in the "bad" example) where an optional parameter (either named or positional) has a non-nullable type and is implicitly initialized to `null`.

So no, this lint won't be necessary under Null Safety because it will be a compilation error.
